autoscheduler/frontend/src/components/LandingPage/SelectTerm/SelectTerm.tsx

```diff
@@ -8,6 +8,23 @@ import { useDispatch } from 'react-redux';
 import setTerm from '../../../redux/actions/term';
 import * as styles from './SelectTerm.css';
 
+/**
+ * Stateful function that caches the json for an api/terms call, so that the API route
+ * only has to be fetched once.
+ */
+const getTermsJson = ((): () => Promise<any> => {
+  let fetched = false;
+  let fetchedData: Promise<any>;
```

Comment:
For personal learning:

So the basic idea is that we're storing the `fetchedData` inside this closure so that we don't need to fetch it again, right? Also, if we were obtaining this data synchronously (not through fetch, maybe like reading from cookies or something), we wouldn't necessarily need to return a Promise, would we?

Reply:
> So the basic idea is that we're storing the `fetchedData` inside this closure so that we don't need to fetch it again, right?
Yeah, it will reuse that same promise every time it's called instead of making a new request
> Also, if we were obtaining this data synchronously (not through fetch, maybe like reading from cookies or something), we wouldn't necessarily need to return a Promise, would we?
Yes, this could work in the same way but with an actual value instead of a promise. A cool thing to note is that since javascript isn't *really* parallel, you don't need any kind of flow control since the entire function will run to completion even if you have some incredibly expensive value to compute, you don't have to worry about race conditions or anything